### ISSUE TITLE
[LLVM-backend] Zig should set `noundef` attributes on arguments

### ISSUE THREAD

**Issue by Validark** [ziglang/zig#18501]
### Zig Version

0.12.0-dev.2076+8fd15c6ca

### Steps to Reproduce and Observed Behavior

[Godbolt Link](https://zig.godbolt.org/z/9EMGoMqMW)
```zig
export fn foo(a: u32, b: u32) bool {
    return a < 2 and b < 2;
}
```
The above is not getting optimized. The obvious optimization here is to transform this into `return (a | b) < 2`, but that will not happen because, according to the fine folks at https://github.com/llvm/llvm-project/issues/77432, Zig is not properly setting the `noundef` attribute on the function arguments, therefore `and` can not be turned into a bitwise AND, i.e. `&`.

### Expected Behavior

It should be as optimized as this code is: ([Godbolt link](https://zig.godbolt.org/z/cqM4jd1h6))

```zig
export fn foo(a: u32, b: u32) bool {
    return 1 == (@intFromBool(a < 2) & @intFromBool(b < 2));
}
```

**Comment by nektro** [comment]
with that attribute does passing `undefined` as a function argument become UB?

**Comment by xdBronch** [comment]
shouldnt passing undefined already be considered UB? it really shouldnt be used like that unless you know the argument is unused

**Comment by nektro** [comment]
yeah I suppose it would. guess I'm wary of what other kind of optimizations it would cause llvm to enable and if we have enough safety checking to warn against it

**Comment by N00byEdge** [comment]
Passing `undefined` as a function argument is perfectly defined, consider for example `syscall6(SYS_EXIT, 0, undefined, undefined, undefined, undefined, undefined)`

**Comment by xdBronch** [comment]
but those parameters are effectively unused right? they're passed to inline asm but only as pseudo values to fullfil the interface of the syscall function. there's no branching on the values for example? I don't mean UB in the sense that the compiler will emit broken code, more that you're willingly using values that may lead to unpredictable behavior. I'm assuming that's more in line with what nektro meant as well

**Comment by N00byEdge** [comment]
Yes, but the LLVM docs don't mention usage, only the values. `If the value representation contains any undefined or poison bits, the behavior is undefined`, that sounds to me like it would cause UB.

**Comment by N00byEdge** [comment]
Ugh, there's something more fucky going on here, and I don't get it.
`v < 2` can be translated to `(v & 0xfffffffe) == 0`.
`(a & 0xfffffffe) == 0 and (b & 0xfffffffe) == 0` should always be equivalent to `((a|b) & 0xfffffffe) == 0`.
This also shows the same issue with zig vs clang, and I'm not sure why. I don't see how undefined bits affect this at all. Is this related to how zig emits IR for the short circuiting on `and`?

**Comment by N00byEdge** [comment]
Yeah, `noundef` has no effect here at all:
```
# llc
define dso_local zeroext i1 @foo(i32 %0, i32 %1) local_unnamed_addr {
Entry:
  %2 = icmp ult i32 %0, 2
  %3 = icmp ult i32 %1, 2
  %4 = select i1 %2, i1 %3, i1 false
  ret i1 %4
}
        .text
        .file   "<stdin>"
        .globl  foo                             # -- Begin function foo
        .p2align        4, 0x90
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:                                # %Entry
        cmpl    $2, %edi
        setb    %cl
        cmpl    $2, %esi
        setb    %al
        andb    %cl, %al
        retq
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
# llc
define dso_local zeroext i1 @foo(i32 noundef %0, i32 noundef %1) local_unnamed_addr {
Entry:
  %2 = icmp ult i32 %0, 2
  %3 = icmp ult i32 %1, 2
  %4 = select i1 %2, i1 %3, i1 false
  ret i1 %4
}
        .text
        .file   "<stdin>"
        .globl  foo                             # -- Begin function foo
        .p2align        4, 0x90
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:                                # %Entry
        cmpl    $2, %edi
        setb    %cl
        cmpl    $2, %esi
        setb    %al
        andb    %cl, %al
        retq
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

**Comment by Plecra** [comment]
Ah here's why: https://llvm.org/docs/LangRef.html#i-select

`select` is a special case: the current version of the function would return `true` with `foo(1, undefined)` because select is well-defined even when the unchosen value is undefined. However the optimized one would ofc return `undef` after the or combined them.

With that cause, I expect another solution would be to implement `and` with bitwise ops instead of select. I don't know if that's preferable :D

**Comment by N00byEdge** [comment]
That makes no sense. Why would select be defined for undefined values?

**Comment by Plecra** [comment]
(By my understanding,) It's currently well-defined to call `foo(1, undefined)` and expect 1 to be returned, which is preventing LLVM from making the optimization called for by the original question. `foo(undefined, *)` is also defined to return an undefined value.

**Comment by N00byEdge** [comment]
Select talks about semantics based on poison bits, not undefined bits. Does zig emit anything that can have poison bits?

**Comment by Plecra** [comment]
https://github.com/llvm/llvm-project/blob/21830c913505b1fd2cf10e454253483180c7e10b/llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp#L4448-L4451

Here's `select`'s behaviour: "`Sa = select Sb, [ (c^d) | Sc | Sd ], [ b ? Sc : Sd ]`" `undef` and `poison` are treated equally, `PoisonValue`s subclass `UndefValue`s.

As a quick explanation of that syntax, `Sx` is the "shadow" of x and contains a bitmask representing which bits of `x` are undefined. `a = b + c` generates `Sa = Sb | Sc` as seen [here](https://github.com/llvm/llvm-project/blob/21830c913505b1fd2cf10e454253483180c7e10b/llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp#L2596-L2601).

*This was the best source I could find for this info: docs are sparse on undef, and most email threads are debating the semantics XD MemorySanitizer at least represents the current intentions of the llvm sources*

**Comment by Plecra** [comment]
It seems the optimization might be valid anyway in this specific case (pointed out by @mlugg on discord). Clang doesn't need it because of their `noundef` annotations, but here's an informal analysis for someone smarter than me to refine 😆 

```
select x, y, x = o < 2^n
where
o = a | b
x = a < 2^n
y = b < 2^n
```
There're 9 cases for each permutation of x and y, and the result value of the instruction is written `r` here. After the rewrite it is `r'`
|value|#1|#2|#3|#4|#5|#6|#7|#8|#9|
|-|-|-|-|-|-|-|-|-|-|
|a |<2^n |<2^n | <2^n | >=2^n |>=2^n |>=2^n |undef |undef |undef|
|b |<2^n |>=2^n| undef |<2^n|  >=2^n |undef |<2^n | >=2^n |undef|
|x |1    |1 |    1     |0     |0|     0     |undef |undef |undef|
|y |1    |0    | undef |1     |0  |   undef |1     |0     |undef|
|r |1    |0    | undef |0     |0  |   0     |undef |undef |undef|
|o |<2^n |>=2^n |_     |>=2^n |>=2^n |>=2^n |_     |>=2^n |_|
|r' |1  |  0|     undef |0     |0     |0|     undef |0     |undef|

It would seem that there're no cases where the rewrite would introduce UB. (I am however entirely out of my element ;))

**Comment by N00byEdge** [comment]
Exactly what I've been sayign about the bitwise rewrite. The undefined bits shouldn't affect this optimization at all. The only thing that could are poison bits.

**Comment by alexrp** [comment]
@mlugg I think you mentioned previously that `noundef` is not appropriate for Zig semantics?
